Data types

viktorvanwijk opened the pull request:
Closes #2324, closes #4883

Note that this [comment](https://github.com/open-formulieren/open-forms/issues/2324#issuecomment-2821603808) contains more issues that need to be solved, but I think it's better to close #2324 with this PR and open new issues for those.

**Changes**

* Cleaned-up methods to get data from the state
* Fixed values of date/time/datetime components configured as multiple not being converted to Python types
* Fixed values of partners component not being converted to Python types
* Removed on-the-fly conversions to Python types
* Added `data_type` and `data_subtype` to `SubmissionValueVariable`

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Release management

  - [x] I have labelled the PR as "needs-backport" accordingly

- I have updated the translations assets (you do NOT need to provide translations)

  - [x] Ran `./bin/makemessages_js.sh`
  - [x] Ran `./bin/compilemessages_js.sh`

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

sergei-maertens commented on docs/developers/backend/core/variables.rst:
```diff
@@ -133,3 +133,21 @@ User defined ``SubmissionValueVariables`` are rendered when the renderer is in m
+Working with submission data
+============================
+
+Accessing the values stored in the ``SubmissionValueVariables`` should done through the
```
```suggestion
Accessing the values stored in the ``SubmissionValueVariables`` should be done through the
```

sergei-maertens commented on docs/developers/backend/core/variables.rst:
```diff
@@ -133,3 +133,21 @@ User defined ``SubmissionValueVariables`` are rendered when the renderer is in m
+values of these components, without having to deal with on-the-fly conversions from (ISO) strings to time-related
+objects.
+
+Note that the data is still stored as JSON in the database, and also exists our Python-type domain once we serialize
```
```suggestion
Note that the data is still stored as JSON in the database, and also exits our Python-type domain once we serialize
```

sergei-maertens commented on src/openforms/formio/utils.py:
```diff
@@ -189,6 +189,23 @@ def get_component_datatype(component):
     return COMPONENT_DATATYPES.get(component_type, FormVariableDataTypes.string)
 
 
+def get_component_data_subtype(component) -> str:
```
```suggestion
def get_component_data_subtype(component: Component) -> str:
```

sergei-maertens commented on src/openforms/forms/models/form_variable.py:
```diff
@@ -292,6 +294,16 @@ class FormVariable(models.Model):
         choices=FormVariableDataTypes.choices,
         max_length=50,
     )
+    data_subtype = models.CharField(
```
Should be have a data migration or `bin` script to set the subtype for form variables that currently have the type `array`?

viktorvanwijk replied:
I was about to suggest that we only need to do this for components configured as multiple, because files and editgrids also have a type array. But this reminds me of something. Our lord and savior editgrid is currently not processed, so any date/time/datetime components that are inside will not be converted :(. I will have to do some more work I think...

sergei-maertens replied:
I :pray: for you

sergei-maertens commented on src/openforms/submissions/models/submission_value_variable.py:
```diff
@@ -327,6 +337,22 @@ class SubmissionValueVariable(models.Model):
         null=True,
         blank=True,
     )
+    data_type = models.CharField(
+        verbose_name=_("data type"),
+        help_text=_("The type of the value that will be associated with this variable"),
+        choices=FormVariableDataTypes.choices,
+        max_length=50,
+    )
+    data_subtype = models.CharField(
+        verbose_name=_("data subtype"),
+        help_text=_(
+            "This field represents the data type of the values inside the container "
+            "for components that are configured as 'multiple'."
+        ),
+        choices=FormVariableDataTypes.choices,
+        max_length=50,
+        blank=True,
+    )
```
same question about data migration here :)

sergei-maertens commented on src/openforms/submissions/models/submission_value_variable.py:
```diff
@@ -359,23 +385,13 @@ def to_python(self, value: VariableValue | object = empty) -> VariableValue:
+        if not self.data_subtype:
+            return self._value_to_python(value, self.data_type)
+        else:
+            assert self.data_type == FormVariableDataTypes.array
```
we *can* express this in a database constraint I think:

```py
condition=(
    (models.Q(data_type=FormVariableDataTypes.array) & ~models.Q(data_subtype=""))
    | (~models.Q(data_type=FormVariableDataTypes.array) & models.Q(data_subtype=""))
)
```

sergei-maertens commented on src/openforms/submissions/models/submission_value_variable.py:
```diff
@@ -419,4 +435,16 @@ def to_python(self, value: VariableValue | object = empty) -> VariableValue:
 
+        if value and data_type == FormVariableDataTypes.partners:
+            # This is a work-around to convert the date of birth string into a date
+            # object. For now, we always expect the ``dateOfBrithPrecision`` to be
```
```suggestion
            # object. For now, we always expect the ``dateOfBirthPrecision`` to be
```

sergei-maertens commented on src/openforms/variables/constants.py:
```diff
@@ -21,6 +21,9 @@ class FormVariableDataTypes(models.TextChoices):
     datetime = "datetime", _("Datetime")
     time = "time", _("Time")
     date = "date", _("Date")
+    # This should only be used as a data subtype, and is a work-around to enable
+    # converting the partners submission data to native Python types
+    partners = "partners", _("Partners")
```
can also be expressed as check constraint!

sergei-maertens commented on src/openforms/variables/utils.py:
```diff
@@ -34,26 +38,18 @@ def check_initial_value(initial_value: JSONObject, data_type: str) -> JSONObject
     if settings.ESCAPE_REGISTRATION_OUTPUT:
-        data = html_escape_json(data)
-
-    formio_data = FormioData(
-        **{
-            variable.key: variable.initial_value
-            for variable in get_static_variables(submission=submission)
-        },
-        **data,
-    )
```
we have a behaviour change here - the static variables are no longer overwritten by submission data. I'm okay with it, but we need to document this like [here](https://open-forms.readthedocs.io/en/3.2.0/changelog.html#upgrade-procedure). Can you add a notice already in the changelog so we don't forget?

sergei-maertens reviewed (commented):
Really, really nice work and I'm pleasantly surprised by the other small cleanups that can be done because `evaluate_form_logic` is now centralizing all that control!

viktorvanwijk commented on src/openforms/forms/migrations/0113_add_missing_data_subtype_to_existing_form_variable.py:
```diff
@@ -0,0 +1,51 @@
+    FormVariable = apps.get_model("forms", "FormVariable")
+    FormDefinition = apps.get_model("forms", "FormDefinition")
+
+    form_definition_iterator = FormDefinition.objects.prefetch_related(
+        "formvariable_set"
+    ).iterator()
```
Ah I see, so basically I'm only loading extra stuff that is not used now...

viktorvanwijk commented:
I combined the data (sub)type migrations where possible, and also fixed some type checking errors that did not occur previously (not entirely sure why). Because there are a lot of changed files, here is a list of files that I changed after the last review
- openforms.registrations.contrib.objects_api.handlers.v2.py
- openforms.registrations.contrib.objects_api.submission_registration.py
- openforms.registrations.contrib.generic_json.plugin.py
